### PR TITLE
[fully_async] fix: drain final dump futures before fit returns

### LINKED ISSUE #7934: [fully_async] fit returns without awaiting or checking its final dump futures
## System Info

- Official `main`: `3efe38c759c14622fd1b2c9e3679f2d02f86bdac` (2026-09-19).
- Affected path: `verl.experimental.fully_async_policy.FullyAsyncTrainer`.
- Reproducer: Python 3.12.13 on macOS arm64, pytest 9.1.1. Source methods are compiled via AST to avoid Ray/model imports. Dump tests use real `ThreadPoolExecutor` workers and temporary local JSONL files; training/validation/checkpoint RPCs are doubles.

## Reproduction

The inherited [`_dump_generations()`](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/trainer/ppo/ray_trainer.py#L473-L512) submits a write to an executor and only checks already-completed futures at submission time. [`FullyAsyncTrainer.fit()`](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/experimental/fully_async_policy/fully_async_trainer.py#L498-L535) overrides the parent training loop, but ends after final validation/checkpointing without calling the existing `_shutdown_dump_executor()`.

A valid interleaving is:

1. The last trainer-owned rollout dump is submitted and is still pending when `_dump_generations()` checks its futures.
2. The queue terminates. No subsequent dump submission checks that future.
3. `fit()` returns successfully while the write is pending, or after the future has completed with an unobserved `OSError`.

This does not mean every dump is lost: the writer may subsequently complete. The bug is the missing completion/error-propagation barrier at the normal return of `fit()`.

[Executable regression test](https://github.com/liruiluo/verl/blob/8f5486999ac3e65a6ca5d0817f910a82352c3a3e/tests/experimental/fully_async_policy/test_async_dump_shutdown_on_cpu.py):

```bash
# Use a new disposable clone; no model dependencies are needed.
git clone --depth 2 --branch fix/fully-async-dump-shutdown https://github.com/liruiluo/verl.git verl-dump-repro
cd verl-dump-repro
uv venv --python 3.12
uv pip install --python .venv/bin/python pytest

# Keep the new regression test, but run the unmodified parent trainer.
git show HEAD^:verl/experimental/fully_async_policy/fully_async_trainer.py > verl/experimental/fully_async_policy/fully_async_trainer.py
.venv/bin/python -B -m pytest -q tests/experimental/fully_async_policy/test_async_dump_shutdown_on_cpu.py

# Restore the proposed fix and rerun the same test.
git restore --source=HEAD -- verl/experimental/fully_async_policy/fully_async_trainer.py
.venv/bin/python -B -m pytest -q tests/experimental/fully_async_policy/test_async_dump_shutdown_on_cpu.py
```

Unmodified trainer: **9 failed, 2 passed**. Proposed fix: **11 passed**. An event holds the real writer while the async test verifies that `fit()` has not returned and the event loop can still run. Releasing the writer either produces a checked JSONL file or raises an injected `OSError` through `fit()`.

Coverage includes rollout-tail writes through all three fit-exit branches (no final validation check, validation check, partial sync), completed but unchecked success/error futures, dumping disabled, and preservation of earlier training/validation exceptions. The validation callback is doubled and does not submit files. Actual validation uses `FullyAsyncRollouter`'s separate executor, which is not drained or tested by this trainer-only patch.

## Expected behavior

After the final validation check and checkpointing, wait for the trainer's inherited dump executor and propagate write errors before returning success. Use `await asyncio.to_thread(self._shutdown_dump_executor)` so the async actor event loop is not blocked by `Future.result()`.

This is a controlled lifecycle/I/O regression test, not evidence of a real disk failure or a model-training effect. The proposed fix concerns the trainer-owned executor on normal termination. The inherited helper propagates its first observed write error; it does not guarantee complete executor cleanup on that error. Rollouter-owned validation dumps, cancellation, abrupt process death and stronger filesystem durability are outside this patch's scope.

## Related work / duplicate check

Searched open and closed issues/PRs for `"_shutdown_dump_executor"`, `fully_async shutdown`, `FullyAsyncTrainer dump`, `dump future`, and `dump async` on 2026-09-19. No open PR adding this normal-exit barrier was found in that search. #6324 introduced asynchronous dump support and #6438 initialized the executor for this subclass. Open #6917 addresses V1 tracking/dataloader cleanup; #6395 handles rollouter failure/EOF propagation. Neither drains this experimental trainer's tail dump futures.

### PR DESCRIPTION
**Pull request by liruiluo**
### What does this PR do?

Fixes #7934.

Await the inherited dump-executor drain at normal `FullyAsyncTrainer.fit()` completion, after the final validation check and checkpointing. Use `asyncio.to_thread` to keep the actor event loop responsive and propagate tail I/O errors.

### Checklist Before Starting

- [x] Search for similar PRs: [area query](https://github.com/verl-project/verl/pulls?q=is%3Apr%20is%3Aopen%20fully_async%20shutdown). [Issue-specific query](https://github.com/verl-project/verl/pulls?q=is%3Apr%20is%3Aopen%207934%20in%3Abody) returned no open PR; the issue and comments were read before submission preparation.
- [x] Use the required PR title format.

Open #6917 concerns V1 tracking/dataloader cleanup; #6395 concerns rollouter errors versus EOF. Neither adds this experimental fit-exit barrier. #6324/#6438 supply the inherited writer and subclass initialization that this patch reuses. No open PR implementing this exact repair was found in the 2026-09-19 scoped searches.

### Test

Python 3.12.13, pytest 9.1.1, macOS arm64:

```bash
python -B -m pytest -q tests/experimental/fully_async_policy/test_async_dump_shutdown_on_cpu.py
pre-commit run --all-files --show-diff-on-failure
```

- 11 passed; unmodified parent with these tests: 9 failed, 2 passed.
- Fresh public fork clone: the same red/green counts, with the final worktree clean.
- Full pre-commit checks: passed.
- Independent code review: passed. Applying both patches together merges without conflicts and passes all 38 targeted tests.
- The new `test_*_on_cpu.py` file is collected by the existing CPU workflow; no workflow change is needed.

Tests execute the real fit and inherited dump methods extracted from source, with real executor threads and temporary local JSONL files. Training, validation and checkpoint RPCs are doubles. Injected I/O errors are not live data-loss evidence. The validation callback is doubled and does not enqueue files; actual validation uses the rollouter's separate executor, outside this patch. The inherited helper reports the first write error without guaranteeing full executor cleanup on error. Cancellation/SIGKILL cleanup, distributed filesystem durability and training quality were not tested.

### API and Usage Example

No API, CLI or config change. Existing users of `experimental/fully_async_policy` receive the correction automatically.

### Design & Code Changes

- A minimal change to the existing fully-async trainer; no V1 migration, reward, GAE, rollout-policy or optimizer change.
- A standalone CPU regression test file, independent of the companion fix.
- Add the current-year Individual Contributor header to the license checker for the new test. The identical two-line support change is also in the companion branch; either merge order is supported.

### Checklist Before Submitting

- [x] Read `CONTRIBUTING.md` and `AGENTS.md`.
- [x] Apply full pre-commit checks (the shared repository hook was already installed).
- [ ] Documentation update: no public API change; the regression test describes the repaired contract.
- [x] Add tests collected by the existing CPU workflow.
- [ ] CI request via the project's channel: not sent.
- [ ] Recipe submodule update: not applicable.

### AI assistance

AI assistance (OpenAI Codex) was used for investigation, patch preparation, test execution and independent review. The human submitter has confirmed completion of the requested review of this exact change before submission. The automated test commands and results reported above were executed and recorded by the assistant.

### CODE REVIEW

**Comment by wuxibin89** [comment]
`verl/experimental/fully_async_policy` is deprecated and will be removed, please V1 trainer
```bash
trainer.use_v1=True
trainer.v1.trainer_mode=separate_async
```
